### PR TITLE
Add QueueV2.CreateQueue and implement it for Cassandra

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
I added the `CreateQueue` method to the `QueueV2` interface. I also implemented it for Cassandra. Finally, I added some validation for the read and write methods of queue messages to check that their corresponding queue exists.

**Why?**
We need to create queues to track the minimum message ID in order to prevent us from scanning over deleted tombstone messages in Cassandra when reading from the front of the queue.

**How did you test it?**
There is 100% test coverage for everything.

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: common/persistence/cassandra/queue_v2_store.go [hunk 1]
```diff
@@ -256,3 +325,53 @@ func (q *queueV2Store) getNextMessageID(ctx context.Context, queueType persisten
...
+				name,
+				queueEncodingStr,
+			),
+		)
+	}
+
+	queue := &persistencespb.Queue{}
+	err = queue.Unmarshal(queueBytes)
+	if err != nil {
+		return nil, serviceerror.NewInternal(
```
Comment (yycptt) [inline common/persistence/cassandra/queue_v2_store.go:366]: `DeserializationError` ?

File: common/persistence/cassandra/queue_v2_store.go [hunk 2]
```diff
@@ -102,6 +104,10 @@ func (q *queueV2Store) EnqueueMessage(
 	request *persistence.InternalEnqueueMessageRequest,
 ) (*persistence.InternalEnqueueMessageResponse, error) {
 	// TODO: add concurrency control around this method to avoid things like QueueMessageIDConflict.
+	_, err := q.getQueue(ctx, request.QueueType, request.QueueName)
```
Comment (yycptt) [inline common/persistence/cassandra/queue_v2_store.go:107]: nit: maybe add another TODO says the existence information can be cached.

File: common/persistence/cassandra/queue_v2_store.go [hunk 3]
```diff
@@ -181,21 +196,75 @@ func (q *queueV2Store) ReadMessages(
...
+	queue *persistencespb.Queue,
+) (int, error) {
+	if len(nextPageToken) == 0 {
+		// Currently, we only have one partition for each queue. However, that might change in the future. If a queue is
+		// created with more than 1 partition by a server on a future release, and then that server is downgraded, we
+		// will need to gracefully handle this case. One way is to return the smallest possible message ID and log a
+		// warning. Since this function is just used as a shortcut to skip over tombstones, all the logic will still be
+		// correct. It just won't be as efficient. Then, the user can upgrade their server to a version which does have
+		// the ability to handle multiple partitions, and the warning will go away.
+		if len(queue.Partitions) != 1 {
```
Comment (yycptt) [inline common/persistence/cassandra/queue_v2_store.go:247]: I think we should just return error here. The read request might be for a different partition and we will always return messages from partition 0 which is very confusing.

File: proto/internal/temporal/server/api/persistence/v1/queues.proto [hunk 4]
```diff
@@ -61,3 +61,20 @@ message HistoryTask {
...
+
+message QueuePartition {
+  // min_message_id is less than or equal to the id of every message in the queue. If there are no messages in the
+  // queue, then this is 0. It is also 0 if there has only been one message enqueued and it has not been deleted yet
+  // because the first message is always assigned id 0. The min_message_id is mainly used to skip over tombstones in
+  // Cassandra: let's say we deleted the first 1K messages from a queue with 1.1K messages. If an operator asked for
+  // the first 100 messages, without the min_message_id, we would have to scan over the 1K tombstone rows before we
+  // could return the 100 messages. With the min_message_id, we can skip over all of the tombstones by specifying
+  // message_id >= queue.min_message_id.
+  int64 min_message_id = 1;
```
Comment (yycptt) [inline proto/internal/temporal/server/api/persistence/v1/queues.proto:74]: same comment as in the other PR re. ranges ideally should be half open. There's no correctness or performance issue though, so I'll leave it to you to decide.

Comment (MichaelSnowden) [reply proto/internal/temporal/server/api/persistence/v1/queues.proto:74]: Hmm, how would it be half-open in this case?

Comment (MichaelSnowden) [reply proto/internal/temporal/server/api/persistence/v1/queues.proto:74]: Discussed offline: we are going to keep this as-is

File: common/util.go [hunk 5]
```diff
@@ -476,7 +476,7 @@ func PrettyPrint[T proto.Message](msgs []T, header ...string) {
 
 // IsValidContext checks that the thrift context is not expired on cancelled.
 // Returns nil if the context is still valid. Otherwise, returns the result of
-// ctx.Err()
+// ctx.EnqueueTaskErr()
```
Comment (yycptt) [inline common/util.go:479]: plz revert

File: service/frontend/http_api_server.go [hunk 6]
```diff
@@ -271,8 +272,8 @@ func (h *HTTPAPIServer) errorHandler(
 	err error,
 ) {
 	// Convert the error using serviceerror. The result does not conform to Google
-	// gRPC status directly (it conforms to gogo gRPC status), but Err() does
-	// based on internal code reading. However, Err() uses Google proto Any
+	// gRPC status directly (it conforms to gogo gRPC status), but EnqueueTaskErr() does
+	// based on internal code reading. However, EnqueueTaskErr() uses Google proto Any
```
Comment (yycptt) [inline service/frontend/http_api_server.go:276]: plz revert

File: common/persistence/cassandra/queue_v2_store.go [hunk 7]
```diff
@@ -256,3 +319,49 @@ func (q *queueV2Store) getNextMessageID(ctx context.Context, queueType persisten
...
+	)
+	if err != nil {
+		if gocql.IsNotFoundError(err) {
+			return nil, persistence.NewQueueNotFoundError(queueType, name)
+		}
+		return nil, gocql.ConvertError("QueueV2GetQueue", err)
+	}
+
+	if queueEncodingStr != enums.ENCODING_TYPE_PROTO3.String() {
+		return nil, serviceerror.NewInternal(
```
Comment (yycptt) [inline common/persistence/cassandra/queue_v2_store.go:347]: nit: UnknownEncodingErr?

**Review by yycptt — APPROVED** [review]

**Pull request merged**